### ISSUE TITLE
"Unhandled rejection HeadlessError: Phantom Process died" when doing multiple open()

### ISSUE THREAD

**Issue by pprmint** [johntitus/node-horseman#45]
Hello! We've been using horseman to access a certain feature in the youtube cms. It was working well before but throws these errors now.

Here's the general logic:
1) We need to login to the youtube cms (no issues what so ever)
2) Loop to different pages and perform some actions (this is where things get bumpy, the error is shown below)

```
horseman .open https://accounts.google.com/ServiceLogin?continue=https%3A%2F%2Fwww.youtube.com%2Fsignin%3Faction_handle_signin%3Dtrue%26next%3D%252F%26app%3Ddesktop%26feature%3Dsign_in_button%26hl%3Den&passive=true&hl=en&service=youtube&uilel=3#identifier +2s
  horseman injected jQuery +13ms
  horseman .click() +2ms
  horseman .type() +10ms input[name="Email"] username@google.com undefined
  horseman .click() +3ms
  horseman .wait() +19ms 2000
  horseman .type() +2s input[name="Passwd"] password123 undefined
  horseman .click() +2ms
  horseman .waitForNextPage() +24ms
  horseman Timeout during waitForNextPage() +5s
  horseman .waitForNextPage() +0ms
  horseman Timeout during waitForNextPage() +5s
  horseman .close(). +1ms
  horseman .open https://www.youtube.com/my_channels?o=<page1> +6ms
  horseman .wait() +0ms 2000
  horseman .open https://www.youtube.com/my_channels?o=<page2> +1ms
Unhandled rejection HeadlessError: Phantom Process died
    at poll_func (/Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:563:10)
    at /Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:61:7
    at ClientRequest.<anonymous> (/Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:484:11)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at Socket.socketErrorListener (_http_client.js:259:9)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at emitErrorNT (net.js:1250:8)
    at doNTCallback2 (node.js:429:9)
    at process._tickCallback (node.js:343:17)
  horseman .open https://www.youtube.com/my_channels?o=<page3> +1ms
Unhandled rejection HeadlessError: Phantom Process died
    at poll_func (/Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:563:10)
    at /Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:61:7
    at ClientRequest.<anonymous> (/Users/ninz/proj/frnky/cms_interface/node_modules/node-horseman/node_modules/node-phantom-simple/node-phantom-simple.js:484:11)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at Socket.socketErrorListener (_http_client.js:259:9)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at emitErrorNT (net.js:1250:8)
    at doNTCallback2 (node.js:429:9)
    at process._tickCallback (node.js:343:17)
  horseman .open https://www.youtube.com/my_channels?o=<pag4> +0ms
```

**Comment by johntitus** [comment]
I'm not sure what's going on off the top of my head, but I notice a `close` statement after you `waitForNextPage`. If you're trying to login/store cookies, I think those would go away after you close it.

Also, you `open` my_channels, then `wait`, then open page 2.  You don't need to `wait` after an `open`, the open will only complete when the page is fully open. 

Also, what version of horseman are you using?

**Comment by pprmint** [comment]
Hello!

Thanks for the reply, im using the latest version of horseman.
Yeah, I noticed the `close` also. Im sure its the cause of the `Phantom Process died`. 

Which brings me to another question, since im looping for the `open()` call then after I opened all pages, I would want to close it. Ill include the snippet of my code so maybe you can understand.
https://gist.github.com/pprmint/7cfea9a66b2813b68fbf

**Comment by johntitus** [comment]
Horseman 2.0 doesn't support tabs yet. Perhaps instead you could open multiple horsemans?

https://github.com/johntitus/node-horseman/blob/master/CHANGELOG.md#200---2015-09-10

**Comment by pprmint** [comment]
I've tried multiple horsemans, but my full implementation of our app forks a worker class that performs the action.

When doing that, there seems to be a conflict with the phantomjs and the forking itself

I will be trying a new approach, ill update you on the outcome

Thanks!

**Comment by johntitus** [comment]
How'd it go @pprmint?

**Comment by pprmint** [comment]
Still no luck.
I've tried opening multiple instance of horseman, and after 6-9 request, phantom died again.
I did not call the `.close()`

**Comment by johntitus** [comment]
I suspect your problem is in the code below.  Not sure how many `tmp_cms` there are, but what this is doing, if I read it right, is telling Phantom to open a new url for each one of them, immediately (not one after the other).  Passing a lot of `opens` to Phantom, before it is done rendering each of them, may cause it to run out of memory, throw a null pointer, who knows.

Phantom has over 1000 issues open so I can't tell you if this is a known problem or not.

```js
for (var tmp_cms in config.CMS) {
            (function(cms_ref) {

                handle.then(function () {
                        return horseman.open('https://www.youtube.com/my_channels?o=' + cms_ref);
                    }).then(function() {
                        return horseman.wait(2000);
                    });
               
            })(config.CMS[tmp_cms]);
        }
```

**Comment by rohitghatol** [comment]
Hi John

I am facing a similar problem, when I try to open and print 2 html pages in parallel, the code either hangs or throws error.

I have setup a git repository where you can try my code and various scenarios first hand and see the hang or errors yourself.

In my code, I am using horseman api's waitFor(fn,value) function, which causes problems related to crash. And if I replace the waitFor(fn,value) with wait(milliseconds), I see a hanging of things.

Everything you need to verify this is mentioned on this github repository - https://github.com/rohitghatol/node-horseman-test-cases

There are 8 test cases over there.

Cheers,
Rohit

**Comment by johntitus** [comment]
@rohitghatol Thank you for the full test case. I'm working through it now, trying to figure out what went wrong.

The first thing I notice, but is not the cause of this problem, is this part:

```js
.click("#Login")
.waitForNextPage()
```

Clicking your login button doesn't actually cause the page to reload, so waitForNextPage is never going to fire. It just times out every time.

I'll keep investigating.

**Comment by johntitus** [comment]
@rohitghatol I found the bug causing your error. I fixed it, and release version 2.2.1.  The `.waitFor` function wasn't closing it's interval variable because I forgot a `var` in front of it.  The issue only caused a problem when multiple horsemen were running at the same time.

It seems to fix your test cases 4 & 6.

@pprmint I'm not sure if this will fix your issue or not. Please grab the latest and try again.  If it's not fixed, please open a new ticket.

**Comment by rohitghatol** [comment]
@johntitus  This worked like a charm. Thank you so much for resolving this in time.

**Comment by pprmint** [comment]
Problem still not fixed :(
Ill probably have to write it in pure phantom and try again.

Thanks for the help though

**Comment by alexey-sh** [comment]
```
        var horseman = new Horseman();
        console.log('get statis for page', url);
        var result = {};

        horseman
            .on('error', function (msg, trace) {
                console.log(msg, trace);
            }).on('timeout', function (timeout, msg) {
                console.log('timeout', msg);
            }).on('resourceTimeout', function (msg) {
                console.log('resourceTimeout', msg);
            }).on('resourceError', function (msg) {
                console.log('resourceError', msg);
            }).on('loadFinished', function (msg) {
                console.log('loadFinished', msg);
            }).on('loadStarted', function (msg) {
                console.log('loadStarted', msg);
            });

        return horseman
            .open(url)
            .evaluate(function (selector) {
                return $(selector).offset();
            }, 'h1')
            .then(function (position) {
                result = extend(result, position);
                return horseman.close();
            }).then(function () {
                console.log('result', result);
                return result;
            });
```

sometimes it works, sometimes it throws an error:
```
horseman .setup() creating phantom instance on 12406 +0ms
get statis for page https://www.quora.com/I-want-to-be-a-billionaire-in-2-years-how-can-I-plan-that
  horseman .open https://www.quora.com/I-want-to-be-a-billionaire-in-2-years-how-can-I-plan-that +13ms
  horseman phantom created. +781ms
  horseman page created +22ms
  horseman .on error set. +3ms
  horseman .on timeout set. +2ms
  horseman .on resourceTimeout set. +0ms
  horseman .on resourceError set. +0ms
  horseman .on loadFinished set. +0ms
  horseman .on loadStarted set. +1ms
  horseman .open: https://www.quora.com/I-want-to-be-a-billionaire-in-2-years-how-can-I-plan-that - status: undefined +30s
Unhandled rejection HeadlessError: Phantom Process died
    at ClientRequest.<anonymous> (/Users/path/node_modules/node-phantom-simple/node-phantom-simple.js:659:12)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at Socket.socketOnEnd (_http_client.js:288:9)
    at emitNone (events.js:72:20)
    at Socket.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:903:12)
    at doNTCallback2 (node.js:439:9)
    at process._tickCallback (node.js:353:17)
```

**Comment by awlayton** [comment]
What version of horseman ad phantom are you using @alexey-sh? Have you tried v3 of horseman?

**Comment by alexey-sh** [comment]
@awlayton 2.8.2

**Comment by awlayton** [comment]
What about your phantom version @alexey-sh?
